token-metadata/js/src/instruction.ts

```diff
@@ -0,0 +1,56 @@
+import type { PublicKey } from '@solana/web3.js';
+
+import { Field, TokenMetadata } from './state';
+
+
+/*
+Not implimented as a strict interface, by function signatures so that the 
+library can work with programs that implement the Rust interface.
+*/
+
+
+// TODO:- Library should impliment Pack/unpack helpers
+export type Pack = (meta: TokenMetadata) => Promise<Buffer>
+export type Unpack = (input: Buffer) => Promise<TokenMetadata>
+
+export type Initialize = (
+    programId: PublicKey,
+    metadata: PublicKey,
+    updateAuthority: PublicKey,
+    mint: PublicKey,
+    mintAuthority: PublicKey,
+    name: string,
+    symbol: string,
+    uri: string,
+) => Promise<void>;
+
+/** If the field does not exist on the account, it will be created.
+ * If the field does exist, it will be overwritten. */
+export type UpdateField = (
+    programId: PublicKey,
+    metadata: PublicKey,
+    updateAuthority: PublicKey,
+    field: Field,
+    value: string
+) => Promise<void>;
```

Comment:
I'm not sure I follow what these types are actually defining.
Rather than a type that takes a defined set of parameters and returns `Promise<void>`, why not just have a set of types for the parameters (pretty much the exact types you have without the arrow function or return bits), and then just define functions to create the instructions?

Something like this:

```typescript
export type UpdateField = {
   /* Type spec */
}

export function createUpdateFieldInstruction(args: UpdateField): TransactionInstruction {
   /* Function body */
}
```

Reply:
I don't think exporting function definitions in the library will work as intended, theres no way to inherit/extend functions in javascript/typescript. 

The intention with these types was that the consumer would write something like the below in their code, and typescript would know arguments and return type:

```
const createUpdateFieldInstruction: UpdateField = (args) => {
  /* Function body*/
}
```

Reply:
But you comment raises another question, I assumed each of the arguments were to be passed individually, but your example has it being passed as a object  - not sure if this is intentional

The different is like this:
1. `createUpdateFieldInstruction(0x000, 0x000, .....)`
2. `createUpdateFieldInstruction({ programId: 0x000, metadata:0x000 })`.


I did 1, because thats what I thought the rust code was doing - but i'm not familiar with rust. Do you know which it is?

I think 2 might be better for us. It would allow updateAuthority to be optional (you can't have optional parameters in the middle of an argument list)

Reply:
> But you comment raises another question, I assumed each of the arguments were to be passed individually, but your example has it being passed as a object - not sure if this is intentional

You're correct, in Rust we're doing individual parameters:

https://github.com/solana-labs/solana-program-library/blob/f76276e91df169d045d5c9c1b240c1923bdd72e3/token-metadata/interface/src/instruction.rs#L225-L236

But it's also not uncommon to pass objects in JS for functions with lots of parameters. I think it can be either one.

Reply:
> I don't think exporting function definitions in the library will work as intended, theres no way to inherit/extend functions in javascript/typescript.

I'm not really sure why you would need to inherit anything from this library, maybe I'm missing something?

All this library needs to do, in general, is provide the ability to build instructions and deserialize state for the SPL Token Metadata interface. When working with a program that implements SPL Token Metadata, like Token2022, this is the library you use to interact with it.

Reply:
Take a look at the latest code - I think its closer to what you had in mind. 

I think I had the wrong understanding about what the library was suppose to be, but its a bit clearer now and should be understanding your comments correctly now!

Reply:
Okay - lets go with object parameters then :)